server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,181 @@
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    private HealthMetadata createHealthMetadata() {
+        return new HealthMetadata(
+            HealthMetadata.DiskThresholds.createDiskThresholds(allocationDiskThresholdSettings, healthDiskThresholdSettings)
+        );
+    }
+
+    private void submitHealthMetadata(String source) {
+        submitHealthMetadata(source, 0);
+    }
+
+    // Visible for testing
+    void submitHealthMetadata(String source, int attempt) {
+        if (attempt < MAX_RETRIES) {
+            HealthMetadata localHealthMedata = createHealthMetadata();
+            String message = attempt == 0 ? source : "retry[" + attempt + "]-" + source;
+            var task = new UpdateHealthMetadataTask(localHealthMedata, () -> submitHealthMetadata(source, attempt + 1));
+            var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+            clusterService.submitStateUpdateTask(message, task, config, taskExecutor);
+        } else {
+            logger.error("Failed to process {} after {} attempts", source, attempt);
+        }
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata, Runnable retry) implements ClusterStateTaskListener {
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "never called";
+        }
+
+        @Override
+        public void onFailure(@Nullable Exception e) {
+            logger.error("unexpected failure during health metadata update", e);
+            retry.run();
+        }
+
+        static class Executor implements ClusterStateTaskExecutor<UpdateHealthMetadataTask> {
+
+            @Override
+            public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateHealthMetadataTask>> taskContexts)
+                throws Exception {
+                final HealthMetadata originalHealthMetadata = HealthMetadata.getHealthCustomMetadata(currentState);
+                HealthMetadata mostRecentHealthMetadata = originalHealthMetadata;
+                if (taskContexts.isEmpty() == false) {
+                    mostRecentHealthMetadata = taskContexts.get(taskContexts.size() - 1).getTask().healthMetadata();
+                    if (Objects.equals(originalHealthMetadata, mostRecentHealthMetadata)) {
+                        mostRecentHealthMetadata = originalHealthMetadata;
+                    }
+                }
+                for (final var taskContext : taskContexts) {
+                    taskContext.success(() -> {});
+                }
```

Comment:
If you move this above the equality check you can make the check just return `currentState`, so something like:

```java
taskContexts.forEach(tc -> tc.success(() -> {}));
if (Objects.equals(originalHealthMetadata, mostRecentHealthMetadata) {
    return currentState;
}
return ClusterState.build(currentState)
    .metadata(Metadata.builder(currentState.metadata()).putCustom(HealthMetadata.TYPE, mostRecentHealthMetadata))
    .build();
```

Reply:
I remodeled it after https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L115.